src/ray/core_worker/task_manager.cc

```diff
@@ -379,59 +379,67 @@ bool TaskManager::RetryTaskIfPossible(const TaskID &task_id) {
     auto it = submissible_tasks_.find(task_id);
     RAY_CHECK(it != submissible_tasks_.end())
-        << "Tried to complete task that was not pending " << task_id;
+        << "Tried to fail task that was not pending " << task_id;
     RAY_CHECK(it->second.pending)
-        << "Tried to complete task that was not pending " << task_id;
+        << "Tried to fail task that was not pending " << task_id;
     spec = it->second.spec;
-    if (!will_retry) {
-      submissible_tasks_.erase(it);
-      num_pending_tasks_--;
-    }
-  }
+    submissible_tasks_.erase(it);
+    num_pending_tasks_--;
 
-  if (!will_retry) {
     // Throttled logging of task failure errors.
-    {
-      absl::MutexLock lock(&mu_);
-      auto debug_str = spec.DebugString();
-      if (debug_str.find("__ray_terminate__") == std::string::npos &&
-          (num_failure_logs_ < kTaskFailureThrottlingThreshold ||
-           (current_time_ms() - last_log_time_ms_) >
-               kTaskFailureLoggingFrequencyMillis)) {
-        if (num_failure_logs_++ == kTaskFailureThrottlingThreshold) {
-          RAY_LOG(WARNING) << "Too many failure logs, throttling to once every "
-                           << kTaskFailureLoggingFrequencyMillis << " millis.";
-        }
-        last_log_time_ms_ = current_time_ms();
-        if (status != nullptr) {
-          RAY_LOG(INFO) << "Task failed: " << *status << ": " << spec.DebugString();
-        } else {
-          RAY_LOG(INFO) << "Task failed: " << spec.DebugString();
-        }
+    auto debug_str = spec.DebugString();
+    if (debug_str.find("__ray_terminate__") == std::string::npos &&
+        (num_failure_logs_ < kTaskFailureThrottlingThreshold ||
+         (current_time_ms() - last_log_time_ms_) > kTaskFailureLoggingFrequencyMillis)) {
+      if (num_failure_logs_++ == kTaskFailureThrottlingThreshold) {
+        RAY_LOG(WARNING) << "Too many failure logs, throttling to once every "
+                         << kTaskFailureLoggingFrequencyMillis << " millis.";
+      }
+      last_log_time_ms_ = current_time_ms();
+      if (status != nullptr) {
+        RAY_LOG(INFO) << "Task failed: " << *status << ": " << spec.DebugString();
+      } else {
+        RAY_LOG(INFO) << "Task failed: " << spec.DebugString();
       }
     }
-    // The worker failed to execute the task, so it cannot be borrowing any
-    // objects.
-    RemoveFinishedTaskReferences(spec, release_lineage, rpc::Address(),
-                                 ReferenceCounter::ReferenceTableProto());
-    if (mark_task_object_failed) {
-      MarkTaskReturnObjectsFailed(spec, error_type, ray_error_info);
-    }
+  }
+
+  // The worker failed to execute the task, so it cannot be borrowing any
+  // objects.
+  RemoveFinishedTaskReferences(spec, true, rpc::Address(),
+                               ReferenceCounter::ReferenceTableProto());
+  if (mark_task_object_failed) {
+    MarkTaskReturnObjectsFailed(spec, error_type, ray_error_info);
+  }
+
+  ShutdownIfNeeded();
```

Comment:
I think this was called regardless of will_retry value. Maybe we should move it to `FailOrRetryPendingTask `?

Reply:
Both `FailOrRetryPendingTask` and `FailPendingTask` call `ShutdownIfNeeded()`, since we may call `FailPendingTask` directly.

Reply:
Is this method idempotent? Now it seems to be possible to call it twice?

Reply:
Do you mind commenting this to `ShutdownIfNeeded ` docstring? (if it wasn't there).